### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/EllipticCurve/DivisionPolynomial/Degree): compute degrees of division polynomials

### PR DESCRIPTION
**Pull request by Multramate**
---

- [x] depends on: #6703 
- [x] depends on: #10814 
- [x] depends on: #10843 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by alreadydone** [comment]
I merged master in [this branch](https://github.com/leanprover-community/mathlib4/compare/EDSRec_merge?expand=1); please merge if it looks good.

**Review by alreadydone — COMMENTED** [review]
I'm trying a new approach [here](https://github.com/leanprover-community/mathlib4/compare/DivisionPolynomial...DivisionPolynomial_degree?expand=1), but I have to sleep now. Maybe you'll finish it when I wake up :)

**Comment by alreadydone** [comment]
Starting working on the new approach again ...

**Comment by Multramate** [comment]
This condenses the eight `natDegree/coeff_even/odd_preΨ'_even/odd` lemmas, but I think we still can't avoid the `natDegree/coeff_preΨ'_add_*` lemmas in the proof of `natDegree_coeff_preΨ'`? Maybe I'm missing something...

**Comment by alreadydone** [comment]
See my [latest commit](https://github.com/leanprover-community/mathlib4/commit/95679ab88fe0a67a2e5cb8e547f1d6eaeb50aae7) for more details.

**Comment by alreadydone** [comment]
Pushed another commit and I think you can handle the rest! Getting back to my own PRs ...

**Comment by alreadydone** [comment]
@Multramate Have you tried to clean up the rest of my branch? If not, I'll try to see if I can golf the subsequent computations today.

**Comment by Multramate** [comment]
> @Multramate Have you tried to clean up the rest of my branch? If not, I'll try to see if I can golf the subsequent computations today.

Thanks for waiting! This is now ready to be reviewed again.

File: Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial/Degree.lean [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+  | one => simpa only [preΨ'_one] using ⟨natDegree_one.le, coeff_one_zero.trans Int.cast_one.symm⟩
+  | two => simpa only [preΨ'_two] using ⟨natDegree_one.le, coeff_one_zero.trans Int.cast_one.symm⟩
+  | three => simpa only [preΨ'_three] using ⟨W.natDegree_Ψ₃, W.coeff_Ψ₃ ▸ (Int.cast_ofNat 3).symm⟩
+  | four =>
+    simpa only [preΨ'_four] using ⟨W.natDegree_preΨ₄, W.coeff_preΨ₄ ▸ (Int.cast_ofNat 2).symm⟩
+  | even m h₁ h₂ h₃ h₄ h₅ =>
+    let dm {m n p q} : _ → _ → (p * q : R[X]).natDegree ≤ m + n := natDegree_mul_le_of_le
+    let dp {m n p} : _ → (p ^ n : R[X]).natDegree ≤ n * m := natDegree_pow_le_of_le n
+    let cm {m n p q} : _ → _ → (p * q : R[X]).coeff (m + n) = _ := coeff_mul_of_natDegree_le
+    let cp {m n p} : _ → (p ^ m : R[X]).coeff (m * n) = _ := coeff_pow_of_natDegree_le
```
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial/Degree.lean:130]: I'm wondering why these are necessary, which aren't there in my branch? I'll check out the branch and take a look myself later ...

Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial/Degree.lean:130]: These shorten the code by a bit! Note that I killed off `natDegree/coeff_even/odd` by doing this.

Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial/Degree.lean:130]: I see, you got rid of `natDegree/coeff_even/odd`! That's really remarkable 😀

**Review by alreadydone — APPROVED** [review]
Thanks, LGTM!
Pushed some trivial golfs
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**